api/base/generic_bulk_views.py

```diff
@@ -0,0 +1,104 @@
+
+from rest_framework import status
+from rest_framework.response import Response
+from rest_framework_bulk import generics as bulk_generics
+from rest_framework.exceptions import PermissionDenied, ValidationError
+
+from framework.auth.core import Auth
+
+from api.base.utils import get_object_or_error
+from api.base.settings import BULK_SETTINGS
+from api.base.exceptions import Conflict
+
+from website.project.model import Node
+
+
+class ListBulkCreateJSONAPIView(bulk_generics.ListBulkCreateAPIView):
+    """
+    Custom ListBulkCreateAPIView that properly formats bulk create responses
+    in accordance with the JSON API spec
+    """
+
+    # overrides ListBulkCreateAPIView
+    def create(self, request, *args, **kwargs):
+        """
+        Correctly formats both bulk and single POST response
+        """
+        response = super(ListBulkCreateJSONAPIView, self).create(request, *args, **kwargs)
+        if 'data' in response.data:
+            return response
+        return Response({'data': response.data}, status=status.HTTP_201_CREATED)
+
+    # overrides ListBulkCreateAPIView
+    def get_serializer(self, *args, **kwargs):
+        """
+        Adds many=True to serializer if bulk operation.
+        """
+
+        if "data" in kwargs:
+            data = kwargs["data"]
+
+            if isinstance(data, list):
+                kwargs['many'] = True
+
+        return super(ListBulkCreateJSONAPIView, self).get_serializer(*args, **kwargs)
+
+
+class BulkUpdateJSONAPIView(bulk_generics.BulkUpdateAPIView):
+    """
+    Custom BulkUpdateAPIView that properly formats bulk update responses in accordance with
+    the JSON API spec
+    """
+
+    # overrides BulkUpdateAPIView
+    def bulk_update(self, request, *args, **kwargs):
+        """
+        Correctly formats bulk PUT/PATCH response
+        """
+        response = super(BulkUpdateJSONAPIView, self).bulk_update(request, *args, **kwargs)
+        return Response({'data': response.data}, status=status.HTTP_200_OK)
+
+
+class BulkDestroyJSONAPIView(bulk_generics.BulkDestroyAPIView):
+    """
+    Custom BulkDestroyAPIView that handles validation and permissions for
+    bulk delete
+    """
+
+    # Overrides BulkDestroyAPIView
+    def bulk_destroy(self, request, *args, **kwargs):
+        """
+        Handles bulk destroy of resource objects.
+
+        Handles some permissions, validation, and enforces bulk limit.
+        """
+        user = self.request.user
+        resource_object_list = []
+        model_cls = kwargs['model']
+        object_type = self.serializer_class.Meta.type_
+
+        if not request.data:
+            raise ValidationError('Request must contain array of resource identifier objects.')
+
+        for item in request.data:
+            item_type = item[u'type']
+            if item_type != object_type:
+                raise Conflict()
+
+            resource_object = get_object_or_error(model_cls, item[u'id'])
```

Comment:
Rather than doing a separate query for each requested object, it'd be more efficient to do a single query, e.g. `Node.find(Q('_id', 'in', requested_ids))`. Perhaps it would make sense to have a `get_objects_or_error` helper.